### PR TITLE
Feature/issue 72 add check to sdm whether data is sorted by design matrix

### PR DESCRIPTION
**Pull request by venpopov**
#### Summary

The sdm is much faster when the data is sorted by the design matrix. This pull request adds a check for the sdmSimple model whether the data is sorted by the design matrix. If not, it gives the user 3 options:
- to quit and sort the data themselves (provides them with code)
- to let bmm sort the data
- to continue with slower estimation

users can control the default behavior via new argument sort_data to fit_model, which can also be set as a global option:
- Null (default): the model asks the user the default query if it detects that the data is unsorted
- True: always sort the data
- False: do not check and do not ask

This check is in a general function order_data_query(), which can be used by any model that would benefit from sorted data.

#### Tests

[x] Confirm that all tests passed
[x] Confirm that devtools::check() produces no errors

#### Release notes

* add a check for the sdmSimple model if the data is sorted by predictors. This leads to much faster sampling. The user can control the default behavior with the `sort_data` argument (#72)

### CODE REVIEW

File: R/utils.R [hunk 1]
```diff
@@ -78,14 +78,23 @@ softmaxinv <- function(p, lambda = 1) {
...
-      opts$chains <-  parallel::detectCores()
+      opts$chains <- parallel::detectCores()
     }
+  } else {
+    cores = NULL
   }
-  exclude <- c("parallel")
+  withr::local_options(
+    list(
+      mc.cores = parallel::detectCores(),
```
Comment (GidonFrischkorn) [inline R/utils.R:90]: mc.cores = cores

**Comment by venpopov** [comment]
This is ready to go

File: DESCRIPTION [hunk 2]
```diff
@@ -1,6 +1,6 @@
 Package: bmm
 Title: Easy and Accesible Bayesian Measurement Models using 'brms'
-Version: 0.3.0.9000
+Version: 0.3.1.9000
```
Comment (GidonFrischkorn) [inline DESCRIPTION:3]: Should we remove the x.x.x.9000 at some point? Or have you left it in there for a specific purpose?

File: R/helpers-data.R [hunk 3]
```diff
@@ -292,4 +293,31 @@ get_standata <- function(formula, data, model, prior=NULL, ...) {
 }
 
 
+# check if the data is sorted by the predictors
+is_data_ordered <- function(data, formula) {
+  predictors <- rhs_vars(formula)
```
Comment (GidonFrischkorn) [inline R/helpers-data.R:298]: I think this needs an additional check if some of the `rhs_vars` appear as `lhs_vars` indicating they are part of a non-linear formula. Probably, you can recycle some code from the `bmf2bf` function here. Would be nice if there is a solution without looping over all formulas.

But something like `dpars <- lhs_vars(formula)` and then `predictors <- predictors[!predictors %in% dpars]` should probably work.

**Review by GidonFrischkorn — COMMENTED** [review]
To accomodate non-linear formulas in the `bmmformula` you probably need to adapt the extraction of the left-hand side arguments and evaluate if some of them also appear as right hand side arguments in the `bmmformula`. I have left a comment for you to check.

File: DESCRIPTION (hunk 2 above)
Comment (venpopov) [reply DESCRIPTION:3]: on the master it doesn't have it. on develop it does. this makes sure that when we make changes to develop, the main website is not updated - rather anything on develop shows up under https://venpopov.github.io/bmm/dev

That way the website is always up to date with the latest release, but we can also see the documentation and changes to the website relevant for the development version

File: R/helpers-data.R (hunk 3 above)
Comment (venpopov) [reply R/helpers-data.R:298]: right, because the predictor then would not be present in the data. good catch

Comment (venpopov) [reply R/helpers-data.R:299]: ok, should be fixed now

**Review by GidonFrischkorn — APPROVED** [review]

**Pull request merged**